### PR TITLE
New Zolertia Re-Mote platform (CC2538 + CC1120)

### PR DESCRIPTION
**Pull request by alignan**
This PR brings the new Zolertia Re-Mote platform (TI CC2538SF53 + TI CC1120), featuring:
* ISM 2.4-GHz IEEE 802.15.4 & Zigbee compliant.
* ISM 868-, 915-, 920-, 950-MHz ISM/SRD Band, IEEE 802.15.4g compliant.
* AES-128/256, SHA2 Hardware Encryption Engine.
* ECC-128/256, RSA Hardware Acceleration Engine for Secure Key Exchange.
* Power consumption down to 3uA using our shutdown mode (disables all hardware but Co-Processor, saves energy of connected sensors and actuators as well).
* Co-Processor to allow peripheral energy management, programming over BSL without requiring to press any button to enter bootloader mode.
* Built-in battery charger (500mA), Energy Harvesting and Solar Panels to be connected to standards LiPo batteries.
* Built-in TMP102 low-power digital temperature sensor (1uA shutdown).
* Power input with wide range 2-26VDC.
* Built-in MicroSD for external storage.
* RGB LED.
* User button mapped to PAx to be able to manually unlock the bootloader.
* 2 x micro USB connectors: USB0 for programming/debugging, USB1 is USB 2.0 ready.
* RF switch to allow programatically to select the 2.4Ghz internal (ceramic chip) or external antenna.
* Small form-factor (as the Z1 mote, half the size of an Arduino) 57x35 mm.

I plan to update the pending features to be ported to RIOT, there's a list of features to be implemented in the platform's [README](https://github.com/alignan/RIOT/tree/pull/remote-port/boards/remote), I can open new PR to be merged on top.

More information at http://zolertia.io/

Cheers!

### CODE REVIEW

**Comment by OlegHahm** [comment]
Awesome!

**Comment by OlegHahm** [comment]
Travis is complaining about some trailing whitespaces.

**Comment by alignan** [comment]
Thanks for the heads-up! I will check those and push -f.

**Comment by OlegHahm** [comment]
Usually commits addressing review issues should be left un-squashed until the review is ready, but in this case, you can squash and force push directly. (I don't want to review a commit with white space removals any way. :wink:)

**Comment by haukepetersen** [comment]
HI, code looks good. I just merged #2852, could you rebase your PR to this new changes (basically it effects the boards Makefiles...). Thx

**Comment by alignan** [comment]
Fixed the whitespaces and changed the Makefile as #2852, squashed and waiting for Travis to behave :-)

**Comment by alignan** [comment]
Updated `Makefile.include`, waiting for Travis again

File: boards/remote/Makefile.include [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+export TERMFLAGS += -p "$(PORT)"
+export DEBUGGER_FLAGS = $(BINDIR) $(ELFFILE)
+export RESET_FLAGS = $(BINDIR)
+export OBJDUMPFLAGS += --disassemble --source --disassembler-options=force-thumb
+
+# setup serial terminal
+include $(RIOTBOARD)/Makefile.include.serial
+
+# this board uses openocd
+include $(RIOTBOARD)/Makefile.include.openocd
```
Comment (haukepetersen) [inline boards/remote/Makefile.include]: If I see it right, you are not using openocd for this board, are you?

File: boards/remote/Makefile.include [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+OS := $(shell uname)
+ifeq ($(OS),Linux)
+  PORT ?= /dev/ttyUSB1
+else ifeq ($(OS),Darwin)
+  PORT ?= $(shell ls -1 /dev/tty.usbserial-* | head -n 1)
+else
+  $(info CAUTION: No flash tool for your host system found!)
+  # TODO: add support for windows as host platform
+endif
+export PORT
```
Comment (haukepetersen) [inline boards/remote/Makefile.include]: You can simplify this section with 
```
PORT_LINUX ?= /dev/ttyUSB1
PORT_DARWIN ?= $(shell ls -1 /dev/tty.usbserial-* | head -n 1)
```
have a look at e.g. the `boards/stm32f4discovery/Makefile.include`...

File: boards/remote/Makefile.include [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+ifeq ($(PROGRAMMER),cc2538-bsl)
+  export FLASHER = python $(RIOTBASE)/dist/tools/cc2538-bsl/cc2538-bsl.py
+  export FFLAGS  = -p "$(PORT)" -e -w -v $(HEXFILE)
+else ifeq ($(PROGRAMMER),jlink)
+  export FLASHER = $(RIOTBOARD)/$(BOARD)/dist/flash.sh
+  export FFLAGS  = $(BINDIR) $(HEXFILE)
+endif
+
+export LINKFLAGS += -L$(RIOTCPU)/$(CPU)
```
Comment (haukepetersen) [inline boards/remote/Makefile.include]: this can be removed also.

File: boards/remote/include/board.h [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+/**
+ * @name LED pin definitions
+ * @{
+ */
+#define LED_PORT_C          GPIO_C
+#define LED_PORT_D          GPIO_D
+#define LED_RED_PIN         (2)
+#define LED_GREEN_PIN       (5)
+#define LED_BLUE_PIN        (3)
+
```
Comment (haukepetersen) [inline boards/remote/include/board.h]: here (and in various other places in this file) are some unnecessary newlines you could remove.

**Comment by haukepetersen** [comment]
looks good. Just some minor things and we should be fine!

**Comment by alignan** [comment]
Fixed the newlines and changed the `Makefile.include`, however I kept the OS check to select either USB port as it is needed for the cc2538 bsl script, renamed the variable to `PORT_BSL`.

**Comment by alignan** [comment]
Mmm I'm confused about the Travis errors, seems unrelated.

**Comment by emmanuelsearch** [comment]
For flashing, it seems to want to use openocd, but then feeds it cc2548-bsl arguments, which fails. 
I get something like the below. I suppose this is not normal. 
(Remark1: I am using OS X). 
(Remark2: I tried your branch some weeks ago, and I did not get this). 


    text	   data	    bss	    dec	    hex	filename
  33616	   2256	   2736	  38608	   96d0	
/path1/hello-world.elf
/path2/openocd.sh -p "/dev/tty.usbserial-0000111DA" -e -w -v /path1/hello-world.bin
Usage: /path2/openocd.sh {flash|debug|debug-server|reset} [extra OpenOCD initialization arguments]

**Comment by alignan** [comment]
Travis is happy again

File: boards/remote/Makefile.include [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# define the cpu used by the Re-mote board
+export CPU       =  cc2538
+export CPU_MODEL ?= cc2538sf53
+
+# define the default port depending on the host OS
+PORT_LINUX  ?= /dev/ttyUSB1
+PORT_DARWIN ?= $(shell ls -1 /dev/tty.usbserial-* | head -n 1)
+
```
Comment (emmanuelsearch) [inline boards/remote/Makefile.include]: I assume that what you mean here is in fact 
```
PORT_DARWIN ?= $(shell ls -1 /dev/tty.usbserial-* | head -n 2 | tail -n 1)
```
right?

Comment (alignan) [inline boards/remote/Makefile.include]: Yes, you are right, thanks for noticing this :-)

Comment (gebart) [inline boards/remote/Makefile.include]: You could use the `wildcard` makefile function for the same purpose, except
it does not depend on the output formatting of any shell commands.

**Comment by emmanuelsearch** [comment]
seems like ipc ping-pong example does not work. Hangs right after the two threads are created. 
Note 1: I compiled/flashed using OS X 10.10.3.
Note 2: during build, I get the following warnings

```
/RIOT/cpu/cc2538/periph/gpio.c: In function ‘isr_gpiob’:
/RIOT/cpu/cc2538/periph/gpio.c:565:50: attention : array subscript is above array bounds [-Warray-bounds]
/RIOT/cpu/cc2538/periph/gpio.c: In function ‘isr_gpioc’:
/RIOT/cpu/cc2538/periph/gpio.c:590:50: attention : array subscript is above array bounds [-Warray-bounds]
/RIOT/cpu/cc2538/periph/gpio.c: In function ‘isr_gpiod’:
/RIOT/cpu/cc2538/periph/gpio.c:615:50: attention : array subscript is above array bounds [-Warray-bounds]
/RIOT/cpu/cortexm_common/thread_arch.c: In function ‘thread_arch_start_threading’:
/RIOT/cpu/cortexm_common/thread_arch.c:264:1: attention : ‘noreturn’ function does return [enabled by default]
```

**Comment by emmanuelsearch** [comment]
Hello World example works (but compile give the same warnings are for the IPC pingpong example mentioned above). Default example compiles, with the same warnings, but the ps command returns nothing.

**Comment by alignan** [comment]
I will give it a second go, the `ps`command was working for me, I will try to get someone with OSX to test as well from our side, thanks for the input!

**Comment by alignan** [comment]
The `default` example works and `ps` seems OK (Ubuntu VM on Windows Host, long story):

ps
2015-06-09 17:34:29,670 - INFO # > ps
2015-06-09 17:34:29,676 - INFO # 	pid | name                 | state    Q | pri | stack ( used) | location   
2015-06-09 17:34:29,682 - INFO # 	  1 | idle                 | pending  Q |  15 |   256 (  136) | 0x20000094 
2015-06-09 17:34:29,689 - INFO # 	  2 | main                 | running  Q |   7 |  1536 (  644) | 0x20000194 
2015-06-09 17:34:29,696 - INFO # 	  3 | uart0                | bl rx    _ |   6 |  1024 (  196) | 0x20000940 
2015-06-09 17:34:29,701 - INFO # 	    | SUM                  |            |     |  2816 (  976)

Is this the expected result?

I will check the IPC ping pong example as well, in the meantime I'm preparing an OSX host to also test.

**Comment by alignan** [comment]
The `ipc_pingpong` example seems to work OK:

(...)
2015-06-09 17:44:25,580 - INFO # 1st: Got msg with content 7424
2015-06-09 17:44:25,581 - INFO # 2nd: Got msg from 2
2015-06-09 17:44:25,583 - INFO # 1st: Got msg with content 7425
2015-06-09 17:44:25,585 - INFO # 2nd: Got msg from 2
2015-06-09 17:44:25,586 - INFO # 1st: Got msg with content 7426
2015-06-09 17:44:25,587 - INFO # 2nd: Got msg from 2
2015-06-09 17:44:25,588 - INFO # 1st: Got msg with content 7427

I will check the OSX.

**Comment by gebart** [comment]
More important: do you get the same out of bounds warnings as Emmanuel?

The effect of out of bounds accesses will vary depending on the memory map
and may therefore seem to work, but after an unrelated change it may stop
working just because the linker placed some variables at a different
location.

**Comment by alignan** [comment]
I got the same warnings:
```
"make" -C /home/zolertia/Desktop/REPO/RIOT/RIOT/cpu/cc2538/periph
/home/zolertia/Desktop/REPO/RIOT/RIOT/cpu/cc2538/periph/gpio.c: In function 'isr_gpiob':
/home/zolertia/Desktop/REPO/RIOT/RIOT/cpu/cc2538/periph/gpio.c:565:50: warning: array subscript is above array bounds [-Warray-bounds]
             state = gpio_config + reverse_pin_lut[GPIO_PXX_TO_NUM(PORT_B, bit)];
                                                  ^
/home/zolertia/Desktop/REPO/RIOT/RIOT/cpu/cc2538/periph/gpio.c: In function 'isr_gpioc':
/home/zolertia/Desktop/REPO/RIOT/RIOT/cpu/cc2538/periph/gpio.c:590:50: warning: array subscript is above array bounds [-Warray-bounds]
             state = gpio_config + reverse_pin_lut[GPIO_PXX_TO_NUM(PORT_C, bit)];
                                                  ^
/home/zolertia/Desktop/REPO/RIOT/RIOT/cpu/cc2538/periph/gpio.c: In function 'isr_gpiod':
/home/zolertia/Desktop/REPO/RIOT/RIOT/cpu/cc2538/periph/gpio.c:615:50: warning: array subscript is above array bounds [-Warray-bounds]
             state = gpio_config + reverse_pin_lut[GPIO_PXX_TO_NUM(PORT_D, bit)];
                                                  ^
"make" -C /home/zolertia/Desktop/REPO/RIOT/RIOT/cpu/cortexm_common
```

Except for this one not showing in my compilation output:

```
/RIOT/cpu/cortexm_common/thread_arch.c: In function ‘thread_arch_start_threading’:
/RIOT/cpu/cortexm_common/thread_arch.c:264:1: attention : ‘noreturn’ function does return [enabled by default]
```

One question, if you press the reset button, does it runs?

**Comment by emmanuelsearch** [comment]
what test are you talking about exactly?
(note: I am currently upgrading Xcode to 7beta, will take a few minutes until I can compile again)

**Comment by alignan** [comment]
Sorry, the icp pingpong.

**Comment by emmanuelsearch** [comment]
Yes, IPC pingpong runs. Here's what I get before it hangs:

```
2015-06-09 16:13:37,066 - INFO # kernel_init(): This is RIOT! (Version: 2014.12-1821-g70e88-MACBOOK-AIR-EB.local-remote-zolertia)
2015-06-09 16:13:37,070 - INFO # kernel_init(): jumping into first task...
2015-06-09 16:13:37,071 - INFO # Starting IPC Ping-pong example...
2015-06-09 16:13:37,071 - INFO # 1st thread started, pid: 2
2015-06-09 16:13:37,072 - INFO # 2nd thread started, pid: 3
```

**Comment by emmanuelsearch** [comment]
back with the latest Xcode etc. Still same warnings and still same issues as described above.

**Comment by gebart** [comment]
Yeah, this is likely a bug in the gpio driver or the configuration header
for the board.

(I don't have any time to assist in debugging something like this until
July, and I don't have the board, so don't assign me)

**Comment by alignan** [comment]
A new update:

* Warnings removed.
* Tested and works on OS X Maverick 10.9.5, 64-bit, tested with arm-none-eabi-gcc 4.9.3 20150303.
* After c286da8 it was required to add the `board.h` to vectors.c as the board wasn't booting.

Currently the `reboot` command is broken for the `cc2538dk` as well, it should be addressed in a separate commit.

**Comment by OlegHahm** [comment]
Kicked Travis.

**Comment by alignan** [comment]
Seems like there are incomplete Travis tests

**Comment by OlegHahm** [comment]
@alignan, rebasing on current master should fix at least some of the Travis failures.

**Comment by OlegHahm** [comment]
```diff
+666
-1
```
I like the diff-stats. :smiling_imp:

**Comment by alignan** [comment]
It wasn't on purpose... maybe... xD

**Comment by alignan** [comment]
Are we good to go?

**Comment by OlegHahm** [comment]
Let's wait until @emmanuelsearch is back next week, since he has the board to test.

**Comment by thomaseichinger** [comment]
@emmanuelsearch ping

**Comment by emmanuelsearch** [comment]
Sorry about the long silence, I was side-tracked. I'm on OS X 10.10.5 and I get the following error when I compile Default, ipc-ping-pong and hello-world:
```
In function ‘thread_arch_start_threading’:
/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/cpu/cortexm_common/thread_arch.c:264:1: erreur: ‘noreturn’ function does return [-Werror]
cc1: all warnings being treated as errors

make[3]: *** [/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/examples/default/bin/remote/cortexm_common/thread_arch.o] Error 1
make[2]: *** [ALL--/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/cpu/cortexm_common] Error 2
make[1]: *** [ALL--/Users/emmanuelB/Documents/work/CODING/RIOT/RIOT/cpu/cc2538] Error 2
make: *** [all] Error 2
``

**Comment by emmanuelsearch** [comment]
Now the good new: with @haukepetersen we verified that Default compiles and works on Linux (Mint16).

**Comment by emmanuelsearch** [comment]
I would say: let's merge this. @alignan can you add to the know issues that Mac OS X is broken for this port?

**Comment by emmanuelsearch** [comment]
ACK

**Comment by haukepetersen** [comment]
Ok, we just decided offline with @emmanuelsearch that we take care about any eventual OSX  changes later, as this PR works like a charm on Linux. ACK

**Comment by alignan** [comment]
Great! I will update the known issues, and in the meantime try to test more with OSX.

**Pull request merged**